### ISSUE TITLE
Plugins with relative paths fail to load when using require hook

### ISSUE THREAD

**Issue by impinball** [babel/babel#2443]
Minimal 100% repro case: https://gist.github.com/impinball/cfe04caff66f7547611a

**Expected:**

```
Success
```

**Found:**

```
$PWD/node_modules/babel/node_modules/babel-core/lib/transformation/file/plugin-manager.js:164
      throw new TypeError(messages.get("pluginNotTransformer", name));
      ^

TypeError: The plugin "./plugin.js" didn't export a Plugin instance
    at PluginManager.validate ($PWD/node_modules/babel/node_modules/babel-core/lib/transformation/file/plugin-manager.js:164:13)
    at PluginManager.add ($PWD/node_modules/babel/node_modules/babel-core/lib/transformation/file/plugin-manager.js:213:10)
    at File.buildTransformers ($PWD/node_modules/babel/node_modules/babel-core/lib/transformation/file/index.js:237:21)
    at new File ($PWD/node_modules/babel/node_modules/babel-core/lib/transformation/file/index.js:139:10)
    at Pipeline.transform ($PWD/node_modules/babel/node_modules/babel-core/lib/transformation/pipeline.js:164:16)
    at Object.transformFileSync ($PWD/node_modules/babel/node_modules/babel-core/lib/api/node.js:137:37)
    at compile ($PWD/node_modules/babel/node_modules/babel-core/lib/api/register/node.js:132:20)
    at normalLoader ($PWD/node_modules/babel/node_modules/babel-core/lib/api/register/node.js:199:14)
    at Object.require.extensions.(anonymous function) [as .js] ($PWD/node_modules/babel/node_modules/babel-core/lib/api/register/node.js:216:7)
    at Module.load (module.js:355:32)
```

@sebmck?

---

**Edit:** See the first comment for a better explanation.

---

The problem appears, from inserting `console.log`s in the generated output, to be caused by [this require](https://github.com/babel/babel/blob/master/packages/babel/src/transformation/file/plugin-manager.js#L82) returning an empty object, as if by a circular require. The catch is that the file is never actually evaluated. And this only happens with relative requires.

It seems to me like the problem is that Babel's require hook is getting in the way of parsing plugins given with relative paths, causing them to be required twice. The best fix IMHO is to explicitly ignore (without the ability to override) plugin files when transpiling.

**Comment by impinball** [comment]
Better explanation: when the plugin is `require`d, Babel's require hook tries to transpile it, resulting in a second, cyclical `require`, as that plugin hasn't even been loaded.

**Comment by impinball** [comment]
And adding the plugin to the `ignore` option does not help. I've done that in the .babelrc

**Comment by jmm** [comment]
@impinball Thanks for the report! As it happens, @brigand and I both independently ran into this yesterday and not realizing it'd already been reported filed a new issue, #2474. So I'm recommending closing that one as a dupe, but how would you feel about linking to @brigand's repro? It's a bit more minimal and there seems to be quite a bit going on in `run.js` here.

I looked at the source and I think Babel doesn't check for `ignore` values from `.babelrc` when the require hook comes to a plugin file.

**Comment by impinball** [comment]
I almost closed this one as @sebmck and I couldn't seem to agree on a solution to this. As for the other issue, go ahead and close it as such.

By the way, the `run.js` is not part of the repro case as much as it's just sugar. I'll clarify in the gist.

**Comment by jmm** [comment]
@impinball Where was that, on slack? This and #2491 are causing me problems that I'm going through some hacky contortions to work around, including using an undocumented option.

> By the way, the `run.js` is not part of the repro case as much as it's just sugar. I removed it from the gist.

Thanks! I think the more minimal probably the better in this case.

**Comment by impinball** [comment]
It was 1-2 days after this bug was filed, IIRC.

**Comment by jmm** [comment]
@impinball Thanks, I found it (same day as issue was filed): https://babeljs.slack.com/archives/general/p1443419353002651. I'm not sure what you guys mean at the end here:

> sebastian [10:48 PM] 
the options are still resolved and parsed
> 
> sebastian [10:49 PM]
> so if you're still getting that "invalid plugin" error when it matches a file in your ignore globs then it's expected
> 
> impinball [10:50 PM] 
> Correct. I just updated the repro gist.

What's the expected part?

You said this:

> impinball [10:46 PM] 
> The fact `ignore` isn't working is a separate bug.

I know you mention the `ignore` part here, but do you consider that part of this issue, or is this issue just pertaining to the idea of automatically treating plugins specially? If the `ignore` part is a separate issue then #2474 can be reopened to track it and this can focus on the other question.

Thanks!

**Comment by impinball** [comment]
Maybe it's a few days later. I remember him stating he'd rather not
introduce more state in the compiler, but I'm going from memory and
replying via phone and email, so it's not the easiest to check at the
moment. I'll try to find the relevant part after I get home.

**Comment by impinball** [comment]
@jmm [Here.](https://babeljs.slack.com/archives/general/p1443493963002733) @sebmck himself said this.

> you're going to need to keep global state
> there's no "this is a plugin file" check anywhere
> i really don't see the point
> ignore your plugin files yourself 

I'm tempted to patch it myself, but I'm not sure if it would be accepted. It's not a common case by any means.

**Comment by jmm** [comment]
@impinball Thanks for that. I read all of that, I'm just not sure what these comments mean:

> sebastian [10:49 PM]
> so if you're still getting that "invalid plugin" error when it matches a file in your ignore globs then it's expected
>
> impinball [10:50 PM] 
> Correct. I just updated the repro gist.

Except for that part ("it's expected") it seems like everyone agrees that including a plugin in `ignore` should prevent the require hook from trying to compile it and triggering the error.

> It's not a common case by any means.

Yeah, I don't know how common it is. It's interesting that you, @brigand, and I all ran into it the same week. It may also become more common once plugins are resolved relative to `.babelrc`'s in v6.

**Comment by impinball** [comment]
> Yeah, I don't know how common it is. It's interesting that you, @brigand, and I all ran into it the same week. It may also become more common once plugins are resolved relative to .babelrc's in v6.

Wow...and currently, they're resolved relative to the current working directory.

**Comment by brigand** [comment]
@sebmck so just to clarify, you're saying that the 'ignore' config option will continue to be ignored for plugins when using the require hook?

**Comment by impinball** [comment]
@sebmck Ping...I think your input is kinda needed here, since there is the chance what you said may have been misinterpreted here. :wink:

**Comment by BerndWessels** [comment]
Hello

What is the status of this?

I still run into this problem. Please see [#2529](https://github.com/babel/babel/issues/2529) for details.

Thank you
Bernd

**Comment by loganfsmyth** [comment]
@BerndWessels It's a circular dependency, so Babel will have to support `ignore`ing it. For now, the fix is to put your transform into another module in `node_modules`. Not ideal, but I believe that is what many people are doing, if they need to use `babel-node`.

**Comment by thejameskyle** [comment]
I don't see how Babel can reliably resolve this circular dependency problem. You're telling a plugin to run on itself, whatever behavior Babel could introduce to get around that would be weird in one way or another.

I'm going to close this as wontfix. Another contributor can reopen it if they want to.

**Comment by loganfsmyth** [comment]
@thejameskyle I think the issue isn't that Babel should resolve the circular dependency, it is that even if you put the file in your `ignore` list, it still doesn't help.

**Comment by thejameskyle** [comment]
Thats not what I'm reading here. Are you able to reproduce this in a repo?

**Comment by loganfsmyth** [comment]
I'm gonna leave this closed in favor of #2474 since it has a clear repro case and focuses on the bug that should be addressed rather than the outward-facing issue.
